core/src/connection/listeners.rs

```diff
@@ -283,12 +285,14 @@ where
                 Poll::Ready(None) => {
                     return Poll::Ready(ListenersEvent::Closed {
                         listener_id: *listener_project.id,
+                        addresses: listener.addresses.to_vec(),
```

Comment:
It's still cloning the content of the container, which is what I was trying to avoid with my suggestion.
```suggestion
                        addresses: listener.addresses.drain(..).collect(),
```

Reply:
Its hard to get good help :)  I did try your suggestion but it didn't seem to work.  Will explore further.  Thanks

Reply:
I've had a bit more of a play with it.  Please correct me if I'm wrong but it seems we cannot get a mutable reference to `listener` (needed if we want to drain `addresses`) because we have already taken a mutable reference.  We would need the following code to be possible (I think, open to correction):
```
    /// Provides an API similar to `Stream`, except that it cannot end.
    pub fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<ListenersEvent<TTrans>> {
        let mut remaining = self.listeners.len();
        while let Some(mut listener) = self.listeners.pop_back() {
            let mut listener_project = listener.as_mut().project();
            let mut addresses = listener.as_mut().addresses;
```
This is clearly not possible.  Is there another way to do this?

Reply:
Ah sorry, you just have to use `listener_project`
```suggestion
                        addresses: listener_project.addresses.drain(..).collect(),
```

Reply:
Cool, thanks.  I learned (a little bit) about pin_project, seems I still have a lot to learn about pinning :) Thanks for your patient revies @tomaka.  One final thing, just to make sure, are you happy with the remaining usage of `SmallVec` in `connection/listener.rs` introduced in this PR?
```
Closed {
         /// The ID of the listener that closed.
         listener_id: ListenerId,
+        /// The addresses that the listener was listening on.
+        addresses: SmallVec<[Multiaddr; 4]>,
```

And, since I'm new and its good to start on the right foot; can you please confirm that the typical pre-PR-push tasks you would expect to pass are only:
- `cargo build`
- `cargo test`

(i.e., no clippy, no e2e tests etc)

Thanks,
Tobin.

Reply:
> are you happy with the remaining usage of SmallVec in connection/listener.rs introduced in this PR?

Ah! No, I missed that. Please turn that into a `Vec` as well.

There are more tests than `cargo test`, but running `cargo test` locally is good enough before opening a PR.

Reply:
Done, I think we are getting dangerously close to a mergeable PR :)

Reply:
Wait. Please revert that last change. Having `SmallVec` in a private field is good.
(you mentioned "introduced in this PR", but your PR didn't touch this.)

Reply:
Oh boy, you are right.  I hope you meant 'revert' figuratively, I removed last patch and force pushed.  Thanks for your patience.